### DISCUSSION TITLE
Android Auto support

### DISCUSSION THREAD

**Discussion by SamuelBrucksch** [mapsforge/vtm#839] [Q&A] [answered]
Hi,

i want to use VTM with Android Auto as Map Renderer. Unfortunately it's not as easy as to add a MapView and everything works. On Android Auto all you get is a `Surface` that provides functions to get a canvas you can draw on:

For example:
```
private  Surface mSurface; // surface is instantiated in SurfaceCallback

void renderFrame() {
  Canvas c = mSurface.lockHardwareCanvas();
  // do some drawing on that canvas
  mSurface.unlockCanvasAndPost(canvas);
}
```

Surface is of type `android.view.Surface` if that helps.

Is there an easy way (maybe something similiar already exists, that i could look at), to use the existing Render logic, to directly draw on a canvas? Or maybe i can also feed the renderer with the Surface and the Renderer does the rest.

Thanks
Samuel

**Comment by SamuelBrucksch** [comment]
This is just a shot in the blue, but could it work to overwrite the onDraw function of `MapView` and copy over the canvas to the AA surface?

Edit: No it does not work. OnDraw is never called. Guess it's not that easy as to just overwrite onDraw().

**Comment by SamuelBrucksch** [comment]
Here is some stripped down sample code of a possible surface renderer from the Android Auto samples:
```
package androidx.car.app.samples.navigation.car;

import android.view.Surface;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.car.app.AppManager;
import androidx.car.app.CarContext;
import androidx.car.app.SurfaceCallback;
import androidx.car.app.SurfaceContainer;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

public final class SurfaceRenderer implements DefaultLifecycleObserver {

    @Nullable
    Surface mSurface;
    @Nullable
    Rect mVisibleArea;
    @Nullable
    Rect mStableArea;

    private final CarContext mCarContext;

    private final SurfaceCallback mSurfaceCallback = new SurfaceCallback() {
        @Override
        public void onSurfaceAvailable(@NonNull SurfaceContainer surfaceContainer) {
            synchronized (SurfaceRenderer.this) {
                Log.i(TAG, String.format("Surface available %s", surfaceContainer));
                mSurface = surfaceContainer.getSurface();
                renderFrame();
            }
        }

        @Override
        public void onVisibleAreaChanged(@NonNull Rect visibleArea) {
            synchronized (SurfaceRenderer.this) {
                Log.i(TAG, String.format("Visible area changed %s. stableArea:%s visibleArea:%s", mSurface, mStableArea,
                        visibleArea));
                mVisibleArea = visibleArea;
                renderFrame();
            }
        }

        @Override
        public void onStableAreaChanged(@NonNull Rect stableArea) {
            synchronized (SurfaceRenderer.this) {
                Log.i(TAG, String.format("Stable area changed %s. stable:%s inset:%s", mSurface, stableArea,
                        mVisibleArea));
                mStableArea = stableArea;
                renderFrame();
            }
        }

        @Override
        public void onSurfaceDestroyed(@NonNull SurfaceContainer surfaceContainer) {
            synchronized (SurfaceRenderer.this) {
                Log.i(TAG, "Surface destroyed");
                mSurface = null;
            }
        }
    };

    public SurfaceRenderer(@NonNull CarContext carContext, @NonNull Lifecycle lifecycle) {
        mCarContext = carContext;
        lifecycle.addObserver(this);
    }

    @Override
    public void onCreate(@NonNull LifecycleOwner owner) {
        Log.i(TAG, "SurfaceRenderer created");
        mCarContext.getCarService(AppManager.class).setSurfaceCallback(mSurfaceCallback);
    }

    void renderFrame() {
        if (mSurface == null || !mSurface.isValid()) {
            // Surface is not available, or has been destroyed, skip this frame.
            return;
        }
        Canvas canvas = mSurface.lockCanvas(null);

        // TODO paint

        mSurface.unlockCanvasAndPost(canvas);
    }
}
```

I would really like to use VTM for this project, as the maps really look awesome, especially in 3D mode. But I can't figure out, how to get it to work with AA and the restriction that we can only draw on a canvas..

**Comment by devemux86** [comment, marked as answer]
> On Android Auto all you get is a Surface that provides functions to get a canvas you can draw on

In such case you could need a new custom [vtm-android](https://github.com/mapsforge/vtm/tree/master/vtm-android) module specifically for Android Auto,
so that the rendering follows its requirements.

VTM (like Mapsforge) has a common graphics API to work on many platforms.
Perhaps Android Auto needs a new module to implement its custom rendering.

**Reply by SamuelBrucksch** [reply]
Hi @devemux86 

thanks for your reply, i somehow expected that. So in general the only thing that differs would be the MapView implementation, right?

https://github.com/mapsforge/vtm/blob/master/vtm-android/src/org/oscim/android/MapView.java

I think i need to adjust the AndroidMap to me more precise:
https://github.com/mapsforge/vtm/blob/c508bde844ca131fc146ff5ae84dd95b2d78d52e/vtm-android/src/org/oscim/android/MapView.java#L218

I already checked all of that, but could not find the part, where the actual drawing happens. If i had found that, it wouldn't be a problem i think. Can you please point me to the exact place, where the map is actually drawn into a view or on a canvas so i can check if i can do the same for AA?

Thanks.

**Comment by devemux86** [comment]
That should be in `MapView.AndroidMap` class.
(there are render and ..frame methods)

And `MapView.GLRenderer` class.

**Reply by SamuelBrucksch** [reply]
Hi,

thanks for providing that information. I found that myself as well, but could not follow where the actual drawing happens. From My understanding the implementation gets a surface somewhere and then draws on it. This might be good in a way, that i just have to find where to pass the surface to draw on. But i can't find that.

How does the GLSurfaceView get the surface to draw on? Can i pass it somewhere? Is it maybe taken from an android view (which i don't have, so i need a way to pass the surface myself).

Thanks for your help.

**Reply by SamuelBrucksch** [reply]
Just to avoid confusion and be a bit more clear what i need:

Android Auto apps are just headless services, that draw on a canvas using a surface and that is then streamed to the car and displayed there. So i have no View or Activity or whatever behind, that does some magic. I really only get a surface and nothing else.

**Reply by ago75** [reply]
Hello,
I'm trying to do the same thing, have you made any progress?
Thank you.

**Reply by SamuelBrucksch** [reply]
There are a few options now, mapbox, here and osmdroid...

**Comment by ago75** [comment]
Thanks for the reply.
I have seen the various alternatives indicated but my idea, having already a complex mobile app with this technology, is to integrate VTM with Android Auto.
Unfortunately my OpenGL programming skills are very limited.
If anyone is interested we could start a collaboration to create a new module 'vtm-android-auto' as suggested by @devemux86.
